registry/manifest.go

```diff
@@ -91,32 +91,45 @@ func (registry *Registry) v1Manifest(repository, reference string, mediaType str
 }
 
 func (registry *Registry) ManifestV2(repository, reference string) (*schema2.DeserializedManifest, error) {
+	deserialized, _, err := registry.ManifestV2WithDigest(repository, reference)
+	return deserialized, err
+}
+
+// ManifestV2WithDigest extends ManifestV2 to return the digest found in the Docker-Content-Digest header.
+// If the header does not exist or is invalid an empty digest is returned (an error is not returned
+// so that existing clients are not affected by new, potentially unrelated, errors).
+func (registry *Registry) ManifestV2WithDigest(repository, reference string) (*schema2.DeserializedManifest, digest.Digest, error) {
 	url := registry.url("/v2/%s/manifests/%s", repository, reference)
 	registry.Logf("registry.manifest.get url=%s repository=%s reference=%s", url, repository, reference)
 
 	req, err := http.NewRequest("GET", url, nil)
 	if err != nil {
-		return nil, err
+		return nil, "", err
 	}
 
 	req.Header.Set("Accept", schema2.MediaTypeManifest)
 	resp, err := registry.Client.Do(req)
 	if err != nil {
-		return nil, err
+		return nil, "", err
 	}
 
 	defer resp.Body.Close()
 	body, err := io.ReadAll(resp.Body)
 	if err != nil {
-		return nil, err
+		return nil, "", err
 	}
 
 	deserialized := &schema2.DeserializedManifest{}
 	err = deserialized.UnmarshalJSON(body)
 	if err != nil {
-		return nil, err
+		return nil, "", err
 	}
-	return deserialized, nil
+
+	digest, err := digest.Parse(resp.Header.Get("Docker-Content-Digest"))
+	if err != nil {
+		digest = ""
```

Comment:
Good idea. It's very possible this header is not sent over:


https://github.com/opencontainers/distribution-spec/blob/main/spec.md#pull
```
A GET request to an existing manifest URL MUST provide the expected manifest, with a response code that MUST be 200 OK. A successful response SHOULD contain the digest of the uploaded blob in the header Docker-Content-Digest.

The Docker-Content-Digest header, if present on the response, returns the canonical digest of the uploaded blob which MAY differ from the provided digest. If the digest does differ, it MAY be the case that the hashing algorithms used do not match. See [Content Digests](https://github.com/opencontainers/image-spec/blob/v1.0.1/descriptor.md#digests) [apdx-3](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#appendix) for information on how to detect the hashing algorithm in use. Most clients MAY ignore the value, but if it is used, the client MUST verify the value against the uploaded blob data.
```

Reply:
The go-containerregistry repo doesn't bother reading this header, as they experienced annoyances with it, though it has been ok for us

Reply:
> The go-containerregistry repo doesn't bother reading this header, as they experienced annoyances with it, though it has been ok for us

That is for sure the better long term solution, which is captured in: ROX-24050

For now, re-using what has been working seems like the 'easy quick fix' to address the known issue until we can allocate the time for the more involved migration / testing.